src/SablierV2MerkleLockupLD.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity >=0.8.22;
+
+import { BitMaps } from "@openzeppelin/contracts/utils/structs/BitMaps.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { ISablierV2LockupDynamic } from "@sablier/v2-core/src/interfaces/ISablierV2LockupDynamic.sol";
+import { Broker, LockupDynamic } from "@sablier/v2-core/src/types/DataTypes.sol";
+import { ud } from "@prb/math/src/UD60x18.sol";
+
+import { SablierV2MerkleLockup } from "./abstracts/SablierV2MerkleLockup.sol";
+import { ISablierV2MerkleLockupLD } from "./interfaces/ISablierV2MerkleLockupLD.sol";
+import { MerkleLockup } from "./types/DataTypes.sol";
+
+/// @title SablierV2MerkleLockupLD
+/// @notice See the documentation in {ISablierV2MerkleLockupLD}.
+contract SablierV2MerkleLockupLD is
+    ISablierV2MerkleLockupLD, // 2 inherited components
+    SablierV2MerkleLockup // 4 inherited components
+{
+    using BitMaps for BitMaps.BitMap;
+    using SafeERC20 for IERC20;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                  STATE VARIABLES
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierV2MerkleLockupLD
+    ISablierV2LockupDynamic public immutable override LOCKUP_DYNAMIC;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                    CONSTRUCTOR
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @dev Constructs the contract by initializing the immutable state variables, and max approving the Sablier
+    /// contract.
+    constructor(
+        MerkleLockup.ConstructorParams memory baseParams,
+        ISablierV2LockupDynamic lockupDynamic
+    )
+        SablierV2MerkleLockup(baseParams)
+    {
+        LOCKUP_DYNAMIC = lockupDynamic;
+
+        // Max approve the Sablier contract to spend funds from the Merkle Lockup contract.
+        ASSET.forceApprove(address(LOCKUP_DYNAMIC), type(uint256).max);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                         USER-FACING NON-CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierV2MerkleLockupLD
+    function claim(
+        uint256 index,
+        address recipient,
+        uint128 amount,
+        LockupDynamic.SegmentWithDuration[] memory segments,
+        bytes32[] calldata merkleProof
+    )
+        external
+        override
+        returns (uint256 streamId)
+    {
+        // Generate the Merkle tree leaf by hashing the corresponding parameters. Hashing twice prevents second
+        // preimage attacks.
+        bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(index, recipient, segments))));
```

Comment:
makes sense. I have changed it to 

```solidity
bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(index, recipient, amount, segments))));
```

Reply:
@andreivladbrg reopening this conversation because I disagree that we should include `amount` in the leaf.

When the protocol fee is non-zero, constructing Merkle tree would require the UI to reverse calculate all the claim amounts to include the protocol fee.

Consider the below simple example
1. Segments sum to 10,000 for each user.
2. If protocol fee = 1%, UI would have to reverse calculate the amount as 10,000/0.9 = 11111 and include it in the Merkle tree. 
3. When a user claims, 11111*0.9 **must** match the segments sum otherwise claim would fail. Now because step 2 requires division in the UI, it is exposed to floating point errors leading to users not being able to claim (since we have strict equality `depositAmount != segmentAmountsSum`). This may require re-submitting the Merkle tree.

So it is safe to not include `amount` in the Merkle leaf and let it be an input parameter into the claim function.